### PR TITLE
Blog: Clean Control Flow in Elixir with Pattern Matching and Immutability

### PR DESCRIPTION
**Pull request by cristineguadelupe**
Create the post file and update the contributors

### CODE REVIEW

**Comment by cristineguadelupe** [comment]
@SophieDeBenedetto Done 😃

**Comment by SophieDeBenedetto** [comment]
I'll review this week and see if we can get a second set of 👀 on here too! 🚀

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
+---
+author: Cristine Guadelupe
+author_link: https://github.com/cristineguadelupe
+categories: general
+date: 2021-01-06
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:5]: We'll want to update this date before we publish 😄

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+layout: post
+title: "Clean Control Flow in Elixir with Pattern Matching and Immutability"
+excerpt: >
+  Learn how to you use pattern matching instead of guard clauses to implement a really clean control flow in Elixir.
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:9]: change `a really clean` to `really clean`

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+---
+One of the features that fascinate me most about Elixir is pattern matching. I always wonder if there is a way to solve what I need using it, and I love exploring it. When you combine the beauty of pattern matching with the power of immutability, some things almost seem magical. But they are not!
+It is not my focus to cover everything about pattern matching and immutability on this post. My goal here is to show you how we can use pattern matching instead of guard clauses to implement a really clean control flow in Elixir.
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:12]: same thing here, `a really clean` -> `really clean`

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+def maybe_attack(player, last_shooter) do
+    player
+    |> make_an_attack()
+    |> set_as_last_shooter()
+end
+```
+
+Wait a second, what have we done here?
+
+In the first `maybe_attack` function we are calling our player (the first argument) as last_shooter!
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:61]: I would rephrase this slightly to read something like "We define the first version of the `maybe_attack` function to take in a first _and_ second argument named `last_shooter`. This means the function will only match if the player provided as a first argument matches the player provided as a second argument". Pretty much same thing you are saying here, but just makes the language a little more clear 😄

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+Our first argument is `player2` and Elixir will bind it with `last_shooter = player2` and since our second argument is also a `last_shooter`, Elixir will use the `^` (pin operator) to check if the previous bind is valid for the second argument instead of trying to rebind it.
+
+```elixir
+last_shooter = player2
+ˆlast_shooter = player1
+```
+
+As player2 is different from player1, we will not have a valid pattern matching and therefore the elixir will move on to try to match with the next function!
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:87]: Change "valid pattern matching" to "valid pattern match"

Comment (SophieDeBenedetto) [reply _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:87]: change "the elixir" to "Elixir"

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+Now the behavior will be different, we are asking Elixir to match two different arguments. That is, to make a bind for each one.
+
+```elixir
+player = player2
+last_shooter = player1
+```
+
+With a valid match, our function will run! Player2 will make an attack and then will be registered as our new last_shooter!
+
+What if player2 tries another attack in a row?
```
Comment (SophieDeBenedetto) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:108]: here, wrap "last_shooter" in backtics "``", same with "maybe_attack" and another mention of "last_shooter" below.

**Review by SophieDeBenedetto — COMMENTED** [review]
This looks great @cristineguadelupe! Very well explained with lots of examples. I left a few comments, mostly minor suggestions to update the wording. Let me know what you think 🚀

**Comment by cristineguadelupe** [comment]
> This looks great @cristineguadelupe! Very well explained with lots of examples. I left a few comments, mostly minor suggestions to update the wording. Let me know what you think 🚀

Thanks for the proofread and suggestions! ❤️

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+
+```elixir
+player = player2
+last_shooter = player1
+```
+
+With a valid match, our function will run! Player2 will make an attack and then will be registered as our new `last_shooter`!
+
+What if player2 tries another attack in a row?
+Well, we will call our first `maybe_attack` function again and try a match. Player2 wants to make an attack and is also `last_shooter`, so we get the following call
```
Comment (doomspork) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:108]:
```suggestion
Well, we will call our first `maybe_attack` function again and try a match. Player2 wants to make an attack and is also `last_shooter`, so we get the following call:
```

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+
+`maybe_attack(player2, player2)`
+
+Trying to match it with the first maybe_attack function we get the following match
```
Comment (doomspork) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:112]:
```suggestion
Trying to match it with the first maybe_attack function we get the following match:
```

File: _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+end
+```
+
+```elixir
+last_shooter = player2
+ˆlast_shooter = player2
+```
+
+Which is a valid match! And since our function just ignores the attack attempt, nothing will happen until another player attempts an attack!
+
```
Comment (doomspork) [inline _posts/2021-06-01-clean-control-flow-in-elixir-with-pattern-matching-and-immutability.md:126]: What do you think about including a couple sentences here to summarize and wrap up the blog post? Something like:

> That's it! We've learned how pattern matching and data immutability together can provide an elegant solution to control flows, another tool in our Elixir toolbox!

**Review by doomspork — CHANGES_REQUESTED** [review]
@cristineguadelupe this is awesome! Like _really_ awesome! I sure hope you'll become a regular contributor to our blog with more great content like this.

Just a few comments but they're fairly minor mostly to do with changing our wording from "you" to "we" in the few places that wasn't already being done.

I did want to ask you and @SophieDeBenedetto your thoughts on perhaps making a few word changes in this to avoid terms that might be upsetting to some given the current atmosphere in the United States. If we could replace `"attack"` with `"move"` and `"shooter"` with `"player"` the concept and game stay the same but use fairly generic game terms. Thoughts?

**Comment by doomspork** [comment]
A few of my comment show "outdated" but they're still applicable, I started these comments before you pushed your changes to those sections for Sophie's feedback 😀

**Comment by cristineguadelupe** [comment]
> @cristineguadelupe this is awesome! Like _really_ awesome! I sure hope you'll become a regular contributor to our blog with more great content like this.
> 
> Just a few comments but they're fairly minor mostly to do with changing our wording from "you" to "we" in the few places that wasn't already being done.
> 
> I did want to ask you and @SophieDeBenedetto your thoughts on perhaps making a few word changes in this to avoid terms that might be upsetting to some given the current atmosphere in the United States. If we could replace `"attack"` with `"move"` and `"shooter"` with `"player"` the concept and game stay the same but use fairly generic game terms. Thoughts?

Hey Sean, thanks a lot for all the feedback and tips. This is my first post and I certainly hope to contribute more frequently! 😃

**Review by SophieDeBenedetto — APPROVED** [review]
This looks good to me! Thanks for addressing all of the comments @cristineguadelupe 😄 I'll let @doomspork take one last look and then we'll update the publish date and get it out 🚀

**Review by doomspork — APPROVED** [review]
This is awesome! Thank you much @cristineguadelupe and @SophieDeBenedetto. I'll let you update the date and merge when ready.

We can get a tweet setup to go out for this too 🚀

**Comment by SophieDeBenedetto** [comment]
> This is awesome! Thank you much @cristineguadelupe and @SophieDeBenedetto. I'll let you update the date and merge when ready.
> 
> We can get a tweet setup to go out for this too 🚀

awesome! Actually looks like the publish date is already set to day 🌞 
I'm going to merge and I'll set up a tweet too!

**Comment by SophieDeBenedetto** [comment]
Post is live here  https://elixirschool.com/blog/clean-control-flow-in-elixir-with-pattern-matching-and-immutability/ with tweet announcement here https://twitter.com/elixirschool/status/1401917723421073413

Thank you @cristineguadelupe 👏 👏 👏
